### PR TITLE
YOLACT -> deepsparse pipeline + annotation script

### PR DESCRIPTION
**Pull request by dbogunowicz**
## Testing:
`python setup.py install`
and then
```
 deepsparse.instance_segmentation.annotate --source src/deepsparse/yolact/sample_images/thailand.jpg
2022-06-30 11:30:33 deepsparse.utils.annotate INFO     Results will be saved to annotation-results/deepsparse-annotations/deepsparse-annotations-017
DeepSparse Engine, Copyright 2021-present / Neuralmagic, Inc. version: 0.13.0.20220609 (ee33a6d7) (release) (optimized) (system=avx2, binary=avx2)
2022-06-30 11:30:44 deepsparse.yolact.annotate INFO     Results saved to annotation-results/deepsparse-annotations/deepsparse-annotations-017
```

image

## Switching from numpy postprocessing to pytorch postprocessing.
Long story short, using numpy together with deepsparse engine is a bit tricky. Numpy likes to grab several cpu threads, so that it hurts the performance of the engine itself. When we restrict numpy to just one thread, the performance of the overall pipeline is by far better. However,  pytorch postprocessing seems not only to be free from that setback (it behaves really nice even when I allow it to grab as many threads as it wants) but is much faster (torch.jit.script functions are being compiled in advance) and finally - could be in future "folded" into the model itself (run in the engine).

image
Note: tested on benchmark server with `num_cores=4`

## A couple of further optimizations of the `annotation` script (outside the pipeline itself):
Here, mostly accelerated the annotation script.
### Mask resizing
The profiler revealed that the `resize` method was taking a lot of time. This was due to the way how I was reshaping the masks so that they have the same size as the annotated image.
`[for cv2.resize(mask, ...) for mask in masks]`
Switching from the list comprehension to single `torch.nn.functional` method gave me **100x** speedup (on this particular transformation).
### Overlaying masks on the image itself
The profiler revealed that the last big bottleneck in the `annotation` function was how I was inpainting the colored mask into the original image. Switching from `numpy.where( mask, color, image)` to the same function, but in PyTorch gave me **10x** speedup (on this particular transformation).
All in all managed to reduce the whole annotation image time **by 4x** (per call = per image).

Before:
image
After: 
image

## Next steps
Created a ticket: https://app.asana.com/0/1201735099598270/1202605644538992/f
to move most of the postprocessing heavy lifting into the model itself (basically, incorporate the torch computational graph for preprocessing into the onnx model).

### CODE REVIEW

File: src/deepsparse/utils/annotate.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def __iter__(self) -> Iterator[Tuple[numpy.ndarray, numpy.ndarray]]:
         self._vid.release()
 
     def __len__(self):
-        return len(self._total_frames)
+        return self._total_frames
```
Comment (dbogunowicz) [inline src/deepsparse/utils/annotate.py:140]: Fixed a tiny bug.

File: src/deepsparse/yolact/annotate.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
```
Comment (dbogunowicz) [inline src/deepsparse/yolact/annotate.py:1]: This is really almost 1-to-1 copy of `src/deepsparse/yolo/annotate.py`

File: src/deepsparse/yolact/utils/annotate.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
```
Comment (dbogunowicz) [inline src/deepsparse/yolact/utils/annotate.py:1]: Very similar to the analogous code for yolo, probably a good idea to merge them.

File: src/deepsparse/yolact/utils/utils.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
```
Comment (dbogunowicz) [inline src/deepsparse/yolact/utils/utils.py:1]: This is a handful of code, but most of that are ports from our yolact repo.

File: src/deepsparse/yolo/annotate.py [hunk 5]
```diff
@@ -81,7 +81,7 @@
 
 
 yolo_v5_default_stub = (
-    "zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/" "pruned-aggressive_96"
+    "zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned-aggressive_96"
```
Comment (dbogunowicz) [inline src/deepsparse/yolo/annotate.py:84]: Boy scout changes

File: src/deepsparse/yolact/pipelines.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+import json
+from typing import Dict, List, Optional, Tuple, Type, Union
+
+import numpy
+
+from deepsparse import Pipeline
+from deepsparse.utils import model_to_path
+from deepsparse.yolact.schemas import YOLACTInputSchema, YOLACTOutputSchema
+from deepsparse.yolact.utils.utils import decode, detect, postprocess, preprocess_array
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/pipelines.py:23]:
```suggestion
from deepsparse.yolact.utils import decode, detect, postprocess, preprocess_array
```

File: src/deepsparse/yolact/utils/annotate.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+# limitations under the License.
+import copy
+from typing import Optional, Tuple
+
+import numpy
+import numpy as np
+
+import cv2
+from deepsparse.yolact.schemas import YOLACTOutputSchema
+from deepsparse.yolo.utils.utils import _get_color, _plot_fps
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/annotate.py:22]:
```suggestion
from deepsparse.yolo.utils import _get_color, _plot_fps
```

File: src/deepsparse/yolact/utils/annotate.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+            y=30,
+            font_scale=0.9,
+            thickness=2,
+        )
+
+    return image_res
+
+
+def _put_mask(
+    image: numpy.ndarray, mask: numpy.ndarray, colour: Tuple[int, int, int]
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/annotate.py:104]: very bad nit, can be absolutely ignored, but we generally use american english in our code base

```suggestion
    image: numpy.ndarray, mask: numpy.ndarray, color: Tuple[int, int, int]
```
Kindly update other usages accordingly if applied

File: src/deepsparse/yolact/utils/annotate.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    )
+
+    image = cv2.putText(
+        image,
+        annotation_text,
+        (int(left), int(top) + text_height),
+        cv2.FONT_HERSHEY_SIMPLEX,
+        0.9,  # font scale
+        (255, 255, 255),  # white text
+        2,  # thickness
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/annotate.py:140]: nit: instead of comments could these be made into variables? like `font_scale=0.9`, then use the variable?

File: src/deepsparse/yolact/utils/annotate.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    )
+    return image
+
+
+def _get_text_size(annotation_text: str) -> Tuple[int, int]:
+    (text_width, text_height), text_baseline = cv2.getTextSize(
+        annotation_text,
+        cv2.FONT_HERSHEY_SIMPLEX,
+        0.9,  # font scale
+        2,  # thickness
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/annotate.py:165]: same as above

File: src/deepsparse/yolact/utils/annotate.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    )
+    text_height += text_baseline
+    return text_width, text_height
+
+
+def _resize_to_fit_img(
+    original_image: numpy.ndarray, masks: numpy.ndarray, boxes
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/annotate.py:172]: annotate type of boxes

File: src/deepsparse/yolact/utils/utils.py [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+    Compute the jaccard overlap of two sets of boxes. The jaccard overlap
+    is simply the intersection over union of two boxes. Here we operate on
+    ground truth boxes and default boxes. If iscrowd=True, put the crowd in box_b.
+    E.g.:
+        A ∩ B / A ∪ B = A ∩ B / (area(A) + area(B) - A ∩ B)
+    Args:
+        box_a: (tensor) Ground truth bounding boxes, Shape: [num_objects,4]
+        box_b: (tensor) Prior boxes from priorbox layers, Shape: [num_priors,4]
+    Return:
+        jaccard overlap: (tensor) Shape: [box_a.size(0), box_b.size(0)]
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/utils.py:85]: Can we be consistent and use `rst` format for docstrings?

File: src/deepsparse/yolact/utils/utils.py [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+
+def sanitize_coordinates(
+    _x1: numpy.ndarray, _x2: numpy.ndarray, img_size: int, padding: int = 0
+) -> Tuple[numpy.ndarray, numpy.ndarray]:
+    """
+    Ported from https://github.com/neuralmagic/yolact/blob/master/layers/box_utils.py
+
+    Sanitizes the input coordinates so that
+    x1 < x2, x1 != x2, x1 >= 0, and x2 <= image_size.
+    Also converts from relative to absolute coordinates.
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/utils.py:116]: nit: add params

File: src/deepsparse/yolact/utils/utils.py [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+def crop(masks: numpy.ndarray, boxes: numpy.ndarray, padding: int = 1) -> numpy.ndarray:
+    """
+    Ported from https://github.com/neuralmagic/yolact/blob/master/layers/box_utils.py
+
+    "Crop" predicted masks by zeroing out everything not in the predicted bbox.
+
+    Args:
+        - masks should be a size [h, w, n] array of masks
+        - boxes should be a size [n, 4] array of bbox coords in relative point form
+    """
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/utils.py:137]: convert to rst

File: src/deepsparse/yolact/utils/utils.py [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+
+    Also note that prior_x and prior_y are center coordinates which
+    is why we have to subtract .5 from sigmoid(pred_x and pred_y).
+
+    Args:
+        - loc:    The predicted bounding boxes of size [num_priors, 4]
+        - priors: The prior box coords with size [num_priors, 4]
+
+    Returns: An array of decoded relative coordinates
+        in point form with size [num_priors, 4]
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/utils.py:273]: Kindly look for other places too where `google` format is used

File: src/deepsparse/yolact/utils/utils.py [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+
+    boxes = numpy.concatenate(
+        (
+            priors[:, :2] + loc[:, :2] * variances[0] * priors[:, 2:],
+            priors[:, 2:] * numpy.exp(loc[:, 2:] * variances[1]),
+        ),
+        1,
+    )
+    boxes[:, :2] -= boxes[:, 2:] / 2
+    boxes[:, 2:] += boxes[:, :2]
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/utils.py:286]: What magic do these lines do?

File: src/deepsparse/yolact/utils/utils.py [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+    nms_threshold: float,
+    max_num_detections,
+    top_k: int,
+) -> Dict[str, numpy.ndarray]:
+    """
+    Ported from
+    https://github.com/neuralmagic/yolact/blob/master/layers/functions/detection.py
+
+    Perform nms for only the max scoring class that isn't background (class 0)
+    """
```
Comment (rahul-tuli) [inline src/deepsparse/yolact/utils/utils.py:305]: params missing

File: src/deepsparse/yolo/pipelines.py [hunk 18]
```diff
@@ -48,10 +50,10 @@
 )
 class YOLOPipeline(Pipeline):
     """
-    Image Segmentation YOLO pipeline for DeepSparse
+    Image detection YOLO pipeline for DeepSparse
```
Comment (rahul-tuli) [inline src/deepsparse/yolo/pipelines.py:53]: Thanks!

File: src/deepsparse/yolo/pipelines.py [hunk 19]
```diff
@@ -243,12 +242,14 @@ def process_engine_outputs(
...
-                ]
-            )
+            batch_labels.append(image_output[:, 5].tolist())
+            if self.class_names is not None:
+                batch_labels[-1] = list(
+                    map(
+                        self.class_names.__getitem__,
+                        list(map(str, list(map(int, batch_labels[-1])))),
+                    )
+                )
```
Comment (rahul-tuli) [inline src/deepsparse/yolo/pipelines.py:252]: This seems overly complicated, can be made simpler by chaining, also it's bad practice to invoke `dunder` methods directly

```suggestion
                batch_labels[-1] = [
                    getattr(self.class_names, str(int(label)))
                    for label in batch_labels[-1]
                ]
```

File: src/deepsparse/yolo/schemas.py [hunk 20]
```diff
@@ -35,11 +35,11 @@
 
 class YOLOInput(BaseModel):
     """
-    Input model for image classification
+    Input model for image detection
```
Comment (rahul-tuli) [inline src/deepsparse/yolo/schemas.py:38]:
```suggestion
    Input model for object detection
```

File: src/deepsparse/yolo/schemas.py [hunk 21]
```diff
@@ -70,7 +70,7 @@ class Config:
 
 class YOLOOutput(BaseModel):
     """
-    Output model for image classification
+    Output model for image detection
```
Comment (rahul-tuli) [inline src/deepsparse/yolo/schemas.py:73]:
```suggestion
    Output model for object detection
```

**Review by rahul-tuli — DISMISSED** [review]

File: src/deepsparse/yolact/schemas.py [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+        description="Confidence threshold applied to the raw detection at "
+        "`detection` step. If a raw detection's score is lower "
+        "than the threshold, it will be automatically discarded",
+    )
+    nms_threshold: float = Field(
+        default=0.5,
+        description="Minimum IoU overlap threshold for a prediction to "
+        "consider it valid (used in Non-Maximum-Suppression step)",
+    )
+    top_k_preprocessing: int = Field(
```
[the remark this answers is not in the document]
Comment (bfineran) [reply src/deepsparse/yolact/schemas.py:56]: +1

**Review by bfineran — COMMENTED** [review]
looks good pending open comments

**Review by bfineran — APPROVED** [review]
approving, however given the current performance, we will need to look into getting these numbers down

File: src/deepsparse/yolact/utils/utils.py [hunk 23]
```diff
@@ -0,0 +1,418 @@
...
+        _x2 = _x2.long()
+    x1 = torch.min(_x1, _x2)
+    x2 = torch.max(_x1, _x2)
+    x1 = torch.clamp(x1 - padding, min=0)
+    x2 = torch.clamp(x2 + padding, max=img_size)
+
+    return x1, x2
+
+
+# @torch.jit.script
```
Comment (KSGulin) [inline src/deepsparse/yolact/utils/utils.py:144]: Is there a reason to keep this comment?

File: src/deepsparse/yolact/annotate.py [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+from deepsparse.yolo.utils.cli_helpers import create_dir_callback
+
+
+yolact_v5_default_stub = (
+    "zoo:cv/segmentation/yolact-darknet53/pytorch/dbolya/coco/pruned82_quant-none"
+)
+
+DEEPSPARSE_ENGINE = "deepsparse"
+ORT_ENGINE = "onnxruntime"
+TORCH_ENGINE = "torch"
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/deepsparse/yolact/annotate.py:88]: I agree with this

**Review by bfineran — APPROVED** [review]
LGTM overall, noting we should add in an eval script that runs the Yolact pipeline (cc @rahul-tuli )

**Comment by bfineran** [comment]
yes

File: src/deepsparse/yolact/annotate.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (kylesayrs) [reply src/deepsparse/yolact/annotate.py:88]: Some of these already exist in `engine.py`

File: src/deepsparse/yolact/annotate.py [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+    is_flag=True,
+    help="Set flag when source is from webcam to not save results."
+    "Not supported for non-webcam sources",
+    show_default=True,
+)
+def main(
+    model_filepath: str,
+    source: str,
+    engine: str,
+    image_shape: tuple,
```
Comment (kylesayrs) [inline src/deepsparse/yolact/annotate.py:175]: nit:
```suggestion
    image_shape: Tuple[int, int],
```

File: src/deepsparse/yolact/pipelines.py [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+        "zoo:cv/segmentation/yolact-darknet53/pytorch/dbolya/coco/pruned82_quant-none"
+    ),
+)
+class YOLACTPipeline(Pipeline):
+    """
+    Image classification pipeline for DeepSparse
+
+    :param model_path: path on local system or SparseZoo stub to load the model from
+    :param engine_type: inference engine to use. Currently, supported values include
+        'deepsparse' and 'onnxruntime'. Default is 'deepsparse'
```
Comment (kylesayrs) [inline src/deepsparse/yolact/pipelines.py:52]: nit: parenthesis in docstring

File: src/deepsparse/yolact/pipelines.py [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+            image_size if isinstance(image_size, Tuple) else (image_size, image_size)
+        )
+        self.top_k = top_k
+
+        super().__init__(**kwargs)
+
+        if isinstance(class_names, str):
+            if class_names.endswith(".json"):
+                class_names = json.load(open(class_names))
+            elif class_names == "coco":
```
Comment (kylesayrs) [inline src/deepsparse/yolact/pipelines.py:90]:
```suggestion
            elif class_names.lower() == "coco":
```

**Review by kylesayrs — DISMISSED** [review]

**Review by kylesayrs — APPROVED** [review]

**Review by corey-nm — APPROVED** [review]
Nice looks great!
